### DISCUSSION TITLE
Dreamer enclosure fan control

### DISCUSSION THREAD

**Discussion by DDMH** [moonglow/FlashForge_Marlin#44] [Q&A] [answered]
Is there a way to enable the enclosure fan control on a Dreamer? A code like M106 P1 S200 no long works. Does the enclosure fan use a different index? 

Alternatively, I can activate the fan by manually set a low enclosure temperature target. Just curious if it still can be done via G-codes.

**Comment by moonglow** [comment, marked as answer]
@DDMH

>A code like M106 P1 S200 no long works

It was never works for my Marlin build ( but works with my patch for original firmware and they are absolutely different firmwares ). Marlin has hardcoded chamber fan to FAN3 ( index 2 ) you can tune it via GUI menu or G-Code. Before try to control it manually you need to disable automatic chamber control ( sets chamber temp to 0 ).

Correct G-Code must be something like that:
`M106 P2 S200`

**Reply by DDMH** [reply]
Thank you. I tested and it worked!

**Comment by DDMH** [comment]
I updated my Flashforge Marlin from v0.13.1 to v0.14.1 today, and the M106/M107 P2 code no longer worked. Later I discovered that the chamber fan was changed from P2 to P1 (fan index 1) in the new firmware. Is it intentional or a bug? Thanks in advance!

**Reply by moonglow** [reply]
Ahh sorry that's my fault, marlin dev team now fixed hardcoded chamber FAN index ( it now can be configured ) so i set it to ( file `pins_FF_MOTHERBOARD.h` )

```C
/* from 2.0.9.3 */
#define FAN1_PIN                            PE1 /* FF rearcase fan */
#define CHAMBER_FAN_INDEX                   1
```

So it is not a bug and now it more correct than before. Thanx you  !

**Reply by DDMH** [reply]
I tested the most recent V0.14.3 version (dreamer_2.0.9.3_dir_swap_la_04242022.bin) and the back fan no longer worked for either temp or manual control. I tried M106 S255 for P1, P2, P3, P4...but there was no response from the back fan.  It's strange because in `pins_FF_MOTHERBOARD.h` I can see that the `CHAMBER_FAN_INDEX` is set as 1.

I restored the firmware to v0.14.1 and M106 P1 S255 worked fine.

**Reply by moonglow** [reply]
Yeah.... you are right... looks like recent updates broke FANs count value. I will fix it ASAP! Thank you!

**Reply by Mrnt** [reply]
@DDMH when you were using the FlashForge firmware on your Dreamer were you able to control the chamber fan using M106? 
The reason I ask is that it appears Marlin expects M106 to only be used for part cooling fans and if you have two fans and two extruders it assumes Fan 1 goes with Extruder 1 Fan 2 with Extruder 2, So if Fan 1 is the part cooling fan and Fan 2 is the chamber fan something like `M106 S255` will turn on the chamber fan rather than the part cooling fan when extruder 2 is active...

**Reply by DDMH** [reply]
@Mrnt Yes, in the v0.14.1 the index `P0` is reserved for the part cooling fan and `P1` for the chamber fan. After setting the chamber fan temp to 0 in the UI (disable automatic chamber fan control), I could control the chamber fan by simply adding `M106 P1 S255` and `M107 P1` to the beginning and the end of a gcode script, respectively. In versions prior to v0.14.1, chamber fan is hardcoded in Marlin as `P2`, and `P1` is seemly unused.

**Reply by Mrnt** [reply]
@DDMH Looks like the Marlin code is kind of screwed up wrt the fans, and I started a bug/discussion about how it is supposed to work on the Marlin repository. 
How would you feel about not controlling the chamber fan directly using M106/M107 but indirectly by setting the target temperature for the chamber using M141 and letting marlin control the fan to try and lower the temperature to the target? The current Flashforge_Marlin implementation is set up to support a heated chamber (even though there is no heat source) and so you could turn the chamber fan by setting the chamber temperature low using M141 or tapping on the enclosure icon on the main interface of the LCD.

**Reply by DDMH** [reply]
@Mrnt Sorry I just found out that you are part of the dev team. Thank you for your contribution!

My dreamer setup is a little different: I have installed a dryer vent on the chamber fans, which is connected to an exhaust grille outside. With M106 and M107, I basically use the chamber fan for fume control instead of temp control. Now I know that I may use M141 to achieve the same goal by using a low chamber temp at the beginning of the gcode and a high temp at the end.
